YJIT: Allocate registers for stack temps

k0kubun opened the pull request:
This PR implements an initial version of register allocation for stack temps.

To make the initial development easy, this version has the following design:

* Do not share the same registers for Out operands and Stack operands.
  *  `get_out_regs` and `get_temp_regs` return distinct registers.
* Spill registers before C calls, jumping to the next block, stack aliasing, and branch stub hit.
  * This operation is inserted as `asm.spill_temps(ctx)`
* For each Assembler, do not allocate registers for indexes that have been spilled before.
  * The number of registers that have been spilled is called `spilled_temps`.
  * ex) If `spilled_temps` is 1, `BP[0]` will use memory and `BP[1]` will use a register. Counted from the stack bottom.

Because it doesn't give significant performance impact yet and it's expected to be somewhat unstable at first, this is disabled by default, and you can enable it by `--yjit-temp-regs=3`.

maximecb commented:
Hi Kokubun. Thank you for working on this and putting this PR together.

Following our pairing last time, I've been thinking about register allocation a bit. There are two things that I would ideally like to see in a register allocator for YJIT.

We have a system that allocates registers for intermediate values inside instructions, and this PR introduces a somewhat distinct system for allocating registers to stack temps that uses a distinct pool of registers. Ideally, both systems should be unified and allocate from the same pool of registers if possible. I know that this may be hard to do, but since registers are a very limited resource, having two distinct pools seems like we might be shooting ourselves in the foot. That being said, I'm open to discussing this. This is not a hard requirement.

The other thing is,  I wonder if this PR is not doing register allocation the wrong way around, so to speak. Registers are allocated to the bottom of the stack. That means if the stack grows to depth > 3, then values at the top of the stack can't be allocated to registers. However, typically, AFAIK, in VM literature, it's the values at the top of the temp stack that you want to allocate to registers for best performance, because these are the most recently pushed values.

Maybe it would make sense, for now, to think of register allocation as something that happens inside of a block. That is, all values must be spilled (or dead, no longer needed) at the end. This frees us from having to map specific stack slots to specific registers.

We often have a pattern where an instruction writes something to the stack, and the next instruction right after reads that value:

```
== BLOCK 3/3, ISEQ RANGE [8,11), 72 bytes =====================
  # regenerate_branch
  # Insn: opt_plus
  # guard arg1 fixnum
  # regenerate_branch
  0x10dbc00a3: test byte ptr [rbx + 8], 1
  0x10dbc00a7: je 0x10dbc20cf
  0x10dbc00ad: mov rax, qword ptr [rbx]
  0x10dbc00b0: sub rax, 1
  0x10dbc00b4: add rax, qword ptr [rbx + 8]
  0x10dbc00b8: jo 0x10dbc20ae
  0x10dbc00be: mov qword ptr [rbx], rax
  # Insn: leave
  # RUBY_VM_CHECK_INTS(ec)
  0x10dbc00c1: mov eax, dword ptr [r12 + 0x20]
  0x10dbc00c6: test eax, eax
  0x10dbc00c8: jne 0x10dbc20e3
  # pop stack frame
  0x10dbc00ce: mov rax, r13
  0x10dbc00d1: add rax, 0x40
  0x10dbc00d5: mov r13, rax
  0x10dbc00d8: mov qword ptr [r12 + 0x10], r13
  0x10dbc00dd: mov rax, qword ptr [rbx]
  0x10dbc00e0: mov rbx, qword ptr [r13 + 8]
  0x10dbc00e4: mov qword ptr [rbx], rax
  0x10dbc00e7: jmp qword ptr [r13 - 8]
```

In the case above, not only do we write then read, but we also write rax to then read into eax, the exact same register. This pattern happens often,  and it seems like it should be fairly trivial to optimize. It almost seems like it could be optimized with a peephole pattern, because it's literally a push followed by a pop.

Alternatively, if we don't want to be limited to just one block, then we could have a system where stack slots are mapped to the register in a cyclic or modulo arithmetic kind of way:

```
slot 0: reg 0
slot 1: reg 1
slot 2: reg 2
slot 3: reg 0
slot 4: reg 1
slot 5: reg 2
slot 6: reg 0
slot 7: reg 1
```

Then, we can keep track of which stack slots are in registers using a bitmap of 8 bits in size.

jimmyhmiller commented:
I probably need to sit with this code a bit longer to understand. But one thing that might help me is an explanation of why we need to spill_temps on splat. I'm sure there is a good reason, but I can't quite think of it right now. Could you maybe explain a bit to help me understand?

k0kubun commented:
It's probably not explained well in the description, but as we discussed in the Google Doc, I intend to break the work into multiple milestones and this is only the first milestone. This is also why it's behind a feature flag that is disabled by default. 

I'm generally more in favor of incrementally developing stuff in master behind a feature flag than growing a big branch and spending a lot of time rebasing it, especially when it doesn't seem too difficult to remove related code as needed.

> We have a system that allocates registers for intermediate values inside instructions, and this PR introduces a somewhat distinct system for allocating registers to stack temps that uses a distinct pool of registers.

What I meant in the PR description was that I do intend to share a set of registers between those two, but in a later version. It still does make sense to split them until you implement more features to it because it makes debugging easier.

> The other thing is, I wonder if this PR is not doing register allocation the wrong way around, so to speak.

Actually this is something I was thinking about after submitting this. At least we could start indexes from the initial stack size. I can try using the stack top as well. I'll have a look at this.

> This pattern happens often, and it seems like it should be fairly trivial to optimize.

Once the operands are converted to registers, I agree that we could do such optimizations in the backend. I think it's the benefit of having the `SpillTemps` IR, because otherwise you can't distinguish where it's safe and where it's not.

> Alternatively, if we don't want to be limited to just one block, then we could have a system where stack slots are mapped to the register in a cyclic or modulo arithmetic kind of way:
> Then, we can keep track of which stack slots are in registers using a bitmap of 8 bits in size.

Interesting idea. When more than 3 stack slots are in use, you also need to track which group of three slots are using registers. But maybe we could just select the first one above the initial stack size.

> I probably need to sit with this code a bit longer to understand. But one thing that might help me is an explanation of why we need to spill_temps on splat.

It's just because the scope of this PR is "Spill registers before jumping to the next block". Spilled slots are read in the callee, which is a different block, so it's spilled. But I do intend to make it not spilled in a later PR.

maximecb commented:
> Actually this is something I was thinking about after submitting this. At least we could start indexes from the initial stack size. I can try using the stack top as well. I'll have a look at this.

I can't remember where I've read this, but I know that caching the top of the stack in a register is a common optimization for stack-based interpreters. Placing the temps at the top of the stack into registers makes sense because they are the most recently used. If you think about it, the temps at the bottom of the stack have been placed there a longer time ago, and the longer a stack temp remains live, the more likely it is that we have to spill it.

> Interesting idea. When more than 3 stack slots are in use, you also need to track which group of three slots are using registers. But maybe we could just select the first one above the initial stack size.

I don't think that there needs to be groups of 3 slots involved. At the moment we track the types for up to 8 temps, so I was thinking with a small bitmap of 8 bits we can independently track which of these up to 8 temps are mapped to registers. This is assuming that each temp, based on its offset from the BP, has a fixed register mapping.

That being said, open to other strategies as well. We could have a system that tries to keep the 2-3 values at the top of the stack in registers instead. What I was suggesting with respect to push and then pop, is that it should be possible, to some extent, to do that in a way that's completely stateless more or less?

> It's probably not explained well in the description, but as we discussed in the Google Doc, I intend to break the work into multiple milestones and this is only the first milestone. This is also why it's behind a feature flag that is disabled by default.
I'm generally more in favor of incrementally developing stuff in master behind a feature flag than growing a big branch and spending a lot of time rebasing it, especially when it doesn't seem too difficult to remove related code as needed.

I understand your concern. However, at this stage, I feel like this is more like a first experiment with register allocation. I think that we need a better strategy, and ideally, we may want to look at metrics. In particular, we may want code size to become smaller. We may want to look at performance on microbenchmarks (30k_ifelse or fib?). We may also want to add a counter to track the number of spills.

I know developing in a branch is not always the most fun, but on the plus side, our backend code seems to have been quite stable for the last few months. I also feel like, although I don't love the current approach, it's impressive that you got it working, and we may not be that far off from a much better solution.

k0kubun commented:
> I can't remember where I've read this, but I know that caching the top of the stack in a register is a common optimization for stack-based interpreters. Placing the temps at the top of the stack into registers makes sense because they are the most recently used. If you think about it, the temps at the bottom of the stack have been placed there a longer time ago, and the longer a stack temp remains live, the more likely it is that we have to spill it.

Currently I have four options in my head:

1. Allocate from the stack bottom (current state)
2. Allocate from the stack bottom, ignoring the initial `spilled_temps` range for an Assembler
3. Allocate from the stack top of an Assembler
4. Allocate from the stack top of an ISEQ

I already gave up on (1). We seem to have better options.

From what I'm hearing, (3) would perform the best. So I prepared a patch, but then I realized that it's hard to make it work. Every Assembler needs to agree on what's the stack_max, so if you calculate stack_max from an Assembler, inline code and outlined code could allocate different registers. But it's not possible to predict what stack_max is gonna look like for inline code when generating a side exit.

A more realistic way to allocate registers from the stack top would be to use ISEQ's `stack_max` (4). When an ISEQ is split in multiple blocks, it might not even use a single register in an Assembler. But most ISEQs might have a shallow stack. Worth checking with a counter.  I'll try this one next.

> This is assuming that each temp, based on its offset from the BP, has a fixed register mapping.

:+1: 

> on the plus side, our backend code seems to have been quite stable for the last few months.

I'd like to point out that codegen.rs changes like splat are also relevant, which has happened recently too. It's reasonable to not merge this until we have some improving metrics in benchmarks though.

k0kubun commented:
I changed it to allocate registers from the stack top https://github.com/ruby/ruby/pull/7430/commits/80372e5fd0f5639d759e6ec83bc71c240a91747b, added counters useful for experiments https://github.com/ruby/ruby/pull/7430/commits/05169daec1fbc2baa3171a42ba563c2992f37635, and increased the number of allocatable registers https://github.com/ruby/ruby/pull/7430/commits/02798e2602c13a8bdea20b549e2c45ae51133e41.

> We may want to look at performance on microbenchmarks (30k_ifelse or fib?)

As of now, the following benchmarks are stably faster:

```
0-regs: ruby 3.3.0dev (2023-03-03T22:44:43Z yjit-stack-temps 02798e2602) +YJIT [x86_64-linux]
6-regs: ruby 3.3.0dev (2023-03-03T22:44:43Z yjit-stack-temps 02798e2602) +YJIT [x86_64-linux]

----------  -----------  ----------  -----------  ----------  -------------  --------------
bench       0-regs (ms)  stddev (%)  6-regs (ms)  stddev (%)  0-regs/6-regs  6-regs 1st itr
30k_ifelse  195.7        0.1         183.6        0.1         1.07           1.00
setivar     7.9          50.6        7.4          53.4        1.07           1.00
----------  -----------  ----------  -----------  ----------  -------------  --------------
```

but I don't see significant impact in other benchmarks yet.

maximecb commented on yjit/src/backend/ir.rs:
```diff
@@ -919,20 +939,48 @@ impl Assembler
 
-    /// Convert Stack operands to memory operands
-    pub fn lower_stack(mut self) -> Assembler
+    /// Allocate registers or memory oprands to Stack operands
```
```suggestion
    /// Allocate registers or memory operands for Stack operands
```

maximecb commented on yjit/src/backend/arm64/mod.rs:
```diff
@@ -171,10 +171,14 @@ impl Assembler
     }
 
+    pub fn get_temp_regs() -> Vec<Reg> {
+        vec![] // FIXME: Not using this yet since sssigning registers to Store insns doesn't work.
```
```suggestion
        vec![] // FIXME: Not using this yet since assigning registers to Store insns doesn't work.
```

maximecb commented:
I think that the way you are handling `asm.spill_temps()` is good :+1:, and it's nice that you are seeing good speedups on the two microbenchmarks.

I see that you've switched to a system where you allocate registers from the top of the stack instead of from the bottom. In my opinion this is still not ideal. We should give priority to the values we just pushed to be allocated to registers, but generally, if we need to spill values, we should spill values at the bottom first. Also, if we use callee-saved registers first, and we have more flexibility as to what we spill, then those callee-saved registers can survive C calls

I think that using an integer index to keep track of which values are spilled is not great. IIUC, in the current scheme, in an ISEQ with `stack_max == 8`, and 6 allocatable registers, the temps at the bottom of the stack can't be mapped to a register. It also seems like the current system is kind of all-or nothing when it comes to spilling.

Right now we only keep track of the types of 8 temp values, and we know that iseqs with more than 8 temps are rare. I would suggest:

- Make it so the assigned register for a temp is `stack_slot_idx % num_temp_regs`
- Use an 8-bit bitmap to track which stack temps are in a register or not

The advantage of this is that, even when the stack is small in an ISEQ with a large `stack_max`, you can still map the stack top to a register. Also, in a situation where a C call would happen, we can spill the caller-save registers, but potentially avoid spilling callee-save registers so that we can keep our stack top in a register. *

Another (maybe less important point) is that ideally, we would allocate both the temps and the out registers from the same pool of registers. That would allow maximum use of those registers.

* The tricky bit here is that we have to be sure that the C function will not GC or side-exit, but it should be possible in some cases.

k0kubun commented:
> We should give priority to the values we just pushed to be allocated to registers, but generally, if we need to spill values, we should spill values at the bottom first.

Ah, I finally see what you've been talking about. I agree that this would be more performant. I'm gonna work on changing it as such next.

> Use an 8-bit bitmap to track which stack temps are in a register or not

I think an 8-bit bitmap alone is not necessarily enough for that scheme when something dynamic like splat or expandarray could suddenly bump SP by 8 slots, but it still seems doable as long as we keep `stack_size` for every operation.

> Another (maybe less important point) is that ideally, we would allocate both the temps and the out registers from the same pool of registers. That would allow maximum use of those registers.

I plan to work on that after exhausting all other optimization ideas in register allocation. Right now, keeping it allows me to read generated code more easily 🙂 

> Also, if we use callee-saved registers first, and we have more flexibility as to what we spill, then those callee-saved registers can survive C calls
> The tricky bit here is that we have to be sure that the C function will not GC or side-exit, but it should be possible in some cases.

It's hard with the IR that this PR currently uses. On C calls, I think we have to represent three different scenarios with IR:

1. A stack temp is known to be an immediate (no need to mark)
2. A stack temp is not known to be an immediate, but a C call is known to not allocate.
3. A stack temp is not known to be an immediate, and a C call is known to allocate.

1 and 2 can skip spilling whereas 3 can't. We can rely on `asm.spill_temps()` to represent "a (next) C call is known to allocate", and simply not generate it for safe C calls. We probably want `Stack::Opnd` to have `immediate: true` to overrule `asm.spill_temps()`.

maximecb commented:
> I think an 8-bit bitmap alone is not necessarily enough for that scheme when something dynamic like splat or expandarray could suddenly bump SP by 8 slots, but it still seems doable as long as we keep stack_size for every operation. Or we could resort to asm.spill_temps() for every such operation if necessary.

Does the current solution avoid spilling temps when expandarray and dynamic splats are used?

> We probably want Stack::Opnd to have immediate: true to overrule asm.spill_temps().

For Higgs, when a value is known to be an immediate, I would directly return an immediate operand. Don't know if that can fit the current design, but if it does, it could help us generate better code.

k0kubun commented:
> Does the current solution avoid spilling temps when expandarray and dynamic splats are used?

I mean they need to be spilled when you run out of registers. So it doesn't, which is the correct behavior. It should keep spilling even after we switch to your rotating allocation strategy, but from lower slots. I'm only saying it's a bit more complicated to spill (not avoid spilling) in that case but should be possible.

> For Higgs, when a value is known to be an immediate, I would directly return an immediate operand. Don't know if that can fit the current design, but if it does, it could help us generate better code.

:+1:

k0kubun commented:
> Use an 8-bit bitmap to track which stack temps are in a register or not

It seems like I was still misunderstanding this part. I assumed the bitmap could be also used for stack slots higher than 8, but I think you didn't mean it since you said "we know that iseqs with more than 8 temps are rare". Never mind about previous replies 🙂 

Anyway, I'm working on the next version. Let's discuss it from there when it's ready.

k0kubun commented:
I changed the allocation state to `[bool; 8]` https://github.com/ruby/ruby/pull/7430/commits/416cce35ca82463de9e533d467ff039247cc6ba6 and used `stack_slot_idx % regs.len()` to choose a register https://github.com/ruby/ruby/pull/7430/commits/12c032a418b5353dab2269b18b86b6a3710fe7e8. This PR is in the state that we were aiming for.

However, because we can't _currently_ bring an allocation state out of an assembler to another, cross-assembler state propagation is still just `u8` (`spilled_size`), which is the stack size on the last `SpillTemps`. We'd need to fix this in order to avoid spilling immediates. This seems would be the current bottleneck for benchmarks using a lot of immediates.

Also, each assembler still keeps increasing `spilled_size` because otherwise it's crashing. We somehow need to fix this without unnecessarily increasing `spilled_size`. This would be the current bottleneck for benchmarks using heap objects.

---

So my milestones for this branch are:

1. Try decreasing `spilled_size` to `stack_size` at each block boundary. A first step towards decreasing `spilled_size`.
2. Mutate an allocation state (`spilled_temps`) as we do `push_insn`, which will allow you to copy the state from inline assembler to outlined assembler for side exits.
4. Propagate `spilled_temps` to outlined assembler instead of `spilled_size`.
5. Modify IR to represent immediate stack operands and avoid spilling such stack operands.
6. Avoid spilling on every block boundary, propagating `spilled_temps` through `Context`.
7. Share the same set of registers between `alloc_temp_regs` and `alloc_out_regs`

maximecb commented on yjit/src/backend/ir.rs:
```diff
@@ -859,19 +867,31 @@ pub struct Assembler
     /// Names of labels
     pub(super) label_names: Vec<String>,
+
+    /// True if each stack slot of the corresponding index is spilled
+    pub(super) spilled_temps: [bool; 8],
```
Does `[bool; 8]` automatically get converted into a `u8` in Rust? If not, might be best to always work with a `u8`. That way it's easier to store in the context without having to translate to/from.

k0kubun replied:
It does:

```rs
    println!("size_of u8: {}", std::mem::size_of::<u8>());
    println!("size_of bool: {}", std::mem::size_of::<bool>());
    println!("size_of [u8; 8]: {}", std::mem::size_of::<[u8; 8]>());
    println!("size_of [bool; 8]: {}", std::mem::size_of::<[bool; 8]>());
```

```
size_of u8: 1
size_of bool: 1
size_of [u8; 8]: 8
size_of [bool; 8]: 8
```

k0kubun replied:
ah well, no. you're right.

k0kubun replied:
It wasn't converted to u8 for the 8 elements. I'll change it to a u8 bitmap.

k0kubun commented:
> It seems like spill_size should no longer be necessary.

In fact it's the part I'm going to work on next 🙂 

> 2. Mutate an allocation state (spilled_temps) as we do push_insn, which will allow you to copy the state from inline assembler to outlined assembler for side exits.
> 3. Propagate spilled_temps to outlined assembler instead of spilled_size.

I believe it still cannot be a bitmap unless you update the bitmap as you do `push_insn`.

k0kubun commented:
> 1. Try decreasing spilled_size to stack_size at each block boundary. A first step towards decreasing spilled_size.

This was a good move. On railsbench,

Before:
```
temp_reg_opnds:                7,974
temp_mem_opnds:               40,080
temp_spills:                   6,980
```

After:
```
temp_reg_opnds:               12,004
temp_mem_opnds:               36,019
temp_spills:                  10,394
```

I'll keep working on the milestones.

maximecb commented:
What is the status on register allocation? How many steps are needed to remove `spilled_size` in favor of always using a small bit map?

k0kubun commented:
I've just finished (2) of https://github.com/ruby/ruby/pull/7430#issuecomment-1461063551. 

I still need to convert `[bool; 8]` to a u8 bitmap, but it's intentional that it's kept as is until it replaces the existing spill size, in order to avoid accidentally mixing up it and `spilled_size` (which is also u8). I'll do that after (3), which will hopefully not be huge effort.

I expect (4) to bring improvements in more benchmarks, and I'm trying to reach that milestone today.

k0kubun commented:
I noticed that one of the rebase commits (https://github.com/ruby/ruby/pull/7430/commits/000e0a0b5d119bf5411aa2a9c7770990e666ed97) started to fail railsbench. It's probably because there's nothing in master that prevents missing `asm.spilled_temps` for new codegen.

Since we decided to not merge that to master, it's probably counterproductive to keep rebasing this branch against master. I'll avoid rebasing this branch until we decide to merge it.

However, leaving conflict means that GitHub Actions doesn't run. I'll move this PR to https://github.com/Shopify/ruby/pull/511 so that we can run CI against an old master.

k0kubun commented:
> How many steps are needed to remove spilled_size in favor of always using a small bit map?

I finished removing `spilled_size` and made it a bit map in https://github.com/Shopify/ruby/pull/511 today. It looks like the base implementation that was passing CI had multiple latent bugs, and the PR hit the second flaky test failure after that. I debugged and fixed it today.

k0kubun commented:
So the next step is to reduce the number of spills. I added stats that count `asm.spill_temps` reasons https://github.com/ruby/ruby/commit/05eeab2ff543b603fdf8a75e66905f60952bfee3. On railsbench, it was:

```
stack temp spill reasons:
    block_end: 57,351,320 (82.0%)
      save_sp:  7,245,873 (10.4%)
       c_args:  3,590,499 ( 5.1%)
        ccall:  1,295,663 ( 1.9%)
    side_exit:    371,628 ( 0.5%)
    iseq_args:     81,769 ( 0.1%)
```

It looks like `block_end` (at the end of each `Block`) is the current bottleneck. If my understanding is correct, we could skip spills at the end of the block as long as we let them survive `branch_stub_hit` calls in the trampoline. I'll work on that next.

maximecb commented:
Great idea to add a spill reason counter 👍
